recipes/dcase2021_task4_separation_SED/local/sed_training.py

```diff
@@ -0,0 +1,485 @@
+                decoded_teacher_strong
+            )
+
+        if torch.any(mask_eval):
+            loss_strong_student = self.supervised_loss(
+                strong_preds_student[mask_eval], labels[mask_eval]
+            )
+            loss_strong_teacher = self.supervised_loss(
+                strong_preds_teacher[mask_eval], labels[mask_eval]
+            )
+
+            output.update(
+                {
+                    "loss_strong_student_on_eval": loss_strong_student,
+                    "loss_strong_teacher_on_eval": loss_strong_teacher,
+                }
+            )
+            filenames_eval = [
+                x
+                for x in filenames
+                if Path(x).parent == Path(self.hparams["data"]["pub_eval_folder"])
+            ]
+
+            decoded_student_strong = batched_decode_preds(
+                strong_preds_student[mask_eval], filenames_eval, self.encoder
+            )
+            self.val_buffer_student_eval = self.val_buffer_student_eval.append(
+                decoded_student_strong
+            )
+
+            decoded_teacher_strong = batched_decode_preds(
+                strong_preds_teacher[mask_eval], filenames_eval, self.encoder
+            )
+            self.val_buffer_teacher_eval = self.val_buffer_teacher_eval.append(
+                decoded_teacher_strong
+            )
+
+        return output
+
+    def validation_epoch_end(self, outputs):
+
+        loss_weak_student_on_weak = nanmean(
+            torch.stack([x["loss_weak_student_on_weak"] for x in outputs])
+        )
+        loss_weak_teacher_on_weak = nanmean(
+            torch.stack([x["loss_weak_teacher_on_weak"] for x in outputs])
+        )
+
+        loss_strong_student_on_synth = nanmean(
+            torch.stack([x["loss_strong_student_on_synth"] for x in outputs])
+        )
+        loss_strong_teacher_on_synth = nanmean(
+            torch.stack([x["loss_strong_teacher_on_synth"] for x in outputs])
+        )
+
+        loss_strong_student_on_eval = nanmean(
+            torch.stack([x["loss_strong_student_on_eval"] for x in outputs])
+        )
+        loss_strong_teacher_on_eval = nanmean(
+            torch.stack([x["loss_strong_teacher_on_eval"] for x in outputs])
+        )
+
+        # TODO maybe use lightning.metrics.f1_score for this
+        # weak dataset
+        ground_truth = pd.read_csv(self.hparams["data"]["weak_val_tsv"], sep="\t")
+        ground_truth = convert_to_event_based(ground_truth)
+
+        save_dir = os.path.join(self.logger.log_dir, "metrics_weak_val")
+        os.makedirs(save_dir, exist_ok=True)
+
+        _, _, weak_student_seg_macro, weak_student_seg_micro = log_sedeval_metrics(
+            self.val_buffer_student_weak, ground_truth, save_dir, self.current_epoch,
+        )
+        _, _, weak_teacher_seg_macro, weak_teacher_seg_micro = log_sedeval_metrics(
+            self.val_buffer_teacher_weak, ground_truth, save_dir, self.current_epoch,
+        )
```

Comment:
The "weak score" is a usual macro f1-score using weak predictions and weak groundtruth.
I don't think this "trick" does the same 🤔 
See : https://github.com/turpaultn/dcase20_task4/blob/dd1d8bb434b08aab6a16a17b8445d4a80a750258/baseline/evaluation_measures.py#L372 
Sorry the code is a bit dirty (it returns the "score per class" so it needs to be averaged)

Reply:
I changed it to standatd F1 macro averaged score over weak labels